src/MongoDB.Driver.Core/Core/Configuration/ConnectionString.cs

```diff
@@ -759,8 +759,17 @@ private void Parse()
                 @"((?<username>[^:@]+)(:(?<password>[^:@]*))?@)?" +
                 serversPattern + @"(/" + databasePattern + ")?/?" + optionsPattern + "$";
 
+            Match badPercentMatch = null;
+            var hexDigit= "0-9A-Fa-f";
+            var percentPattern = $"(%[^{hexDigit}])|"  // % followed by a non-hexadecimal digit OR
+                                 +$"(%[{hexDigit}][^{hexDigit}])"; // % followed by a hexadecimal digit and a non-hexadecimal digit
+            if (_originalConnectionString.Contains("%"))
+            {
+                badPercentMatch = Regex.Match(_originalConnectionString, percentPattern);
+            }
+
```

Comment:
I would not mix the new validation (for valid % escape sequences) with the existing validation (for overall structure). Perhaps something like this at this location in the file:

    var invalidPercentPattern = @"%$|%.$|%[^0-9a-fA-F]|%[0-9a-fA-F][^0-9a-fA-F]";
    if (Regex.IsMatch(_originalConnectionString, invalidPercentPattern))
    {
        var message = string.Format("The connection string '{0}' contains an invalid '%' escape sequence.", _originalConnectionString);
        throw new MongoConfigurationException(message);
    }

I've added a few cases to the regular expression to cover percent signs near the end of the string that are not followed by the required two characters.

Reply:
p.s. It's possible that an approach not using regular expressions might be simpler but maybe not.

Reply:
@vincentkam and I tried to validate the connection string using the library function `Uri.IsWellFormedUriString(String, UriKind)` but that doesn't work with pre-existing test cases, so I think regular expressions may be our best bet.

Reply:
That gave me an idea. The following helper method probably performs much faster than regular expressions:

    public static bool AllPercentEncodingsAreValid(string value)
    {
        for (var i = 0; i < value.Length; i++)
        {
            if (value[i] == '%' && !Uri.IsHexEncoding(value, 0))
            {
                return false;
            }
        }

        return true;
    }

Or just to tease @vincentkam here's a slower functional version:

    public static bool AllPercentEncodingsAreValid2(string value)
        => Enumerable.Range(0, value.Length).All(i => value[i] != '%' || Uri.IsHexEncoding(value, i));

Reply:
LOL at the functional version tease 😸 The docs do state that `All()` returns early so it should only be a tiny tiny bit slower due to the overhead of LINQ.

Also doable would be...
`s.Select((_, i) => i).All(i => s[i] != '%' || Uri.IsHexEncoding(s, i))` 😆

Reply:
After offline discussion with @vincentkam, I decided to use an #ifdef for `AllPercentEncodingsAreValid(...)`, since the function [`Uri.IsHexEncoding(...)`](https://docs.microsoft.com/en-us/dotnet/api/system.uri.ishexencoding?view=netframework-4.8#System_Uri_IsHexEncoding_System_String_System_Int32_) is not available in .NETStandard 1.5, which is what we're targeting. I kept the regex way as an alternative. We could also choose to vendor in the library function if that would make more sense.